core/src/rpc.rs

```diff
@@ -1299,6 +1292,103 @@ impl JsonRpcRequestProcessor {
+    fn get_filtered_spl_token_accounts_by_owner(
+        &self,
+        bank: &Arc<Bank>,
+        owner_key: &Pubkey,
+        mut filters: Vec<RpcFilterType>,
+    ) -> Vec<(Pubkey, Account)> {
+        // The by-owner accounts index checks for Token Account state and Owner address on inclusion.
+        // However, due to the current AppendVec implementation, accounts may remain in the index after
+        // being wiped and reinitialized as something else. We include the redundant filters here to
+        // avoid returning these accounts.
+        //
+        // Filter on Token Account state
+        filters.push(RpcFilterType::DataSize(
+            TokenAccount::get_packed_len() as u64
+        ));
+        // Filter on Owner address
+        filters.push(RpcFilterType::Memcmp(Memcmp {
+            offset: 32,
+            bytes: MemcmpEncodedBytes::Binary(owner_key.to_string()),
+            encoding: None,
+        }));
+
+        if self
+            .config
+            .supported_indexes
+            .contains(&IndexType::TokenOwner)
+        {
+            bank.get_filtered_indexed_accounts(&IndexKey::TokenOwner(*owner_key), |account| {
+                filters.iter().all(|filter_type| match filter_type {
+                    RpcFilterType::DataSize(size) => account.data.len() as u64 == *size,
+                    RpcFilterType::Memcmp(compare) => compare.bytes_match(&account.data),
+                })
+            })
+        } else {
+            self.get_filtered_program_accounts(bank, &spl_token_id_v2_0(), filters)
+        }
+    }
+
+    /// Get an iterator of spl-token accounts by mint address
+    fn get_filtered_spl_token_accounts_by_mint(
+        &self,
+        bank: &Arc<Bank>,
+        mint_key: &Pubkey,
+        mut filters: Vec<RpcFilterType>,
+    ) -> Vec<(Pubkey, Account)> {
+        // The by-mint accounts index checks for Token Account state and Mint address on inclusion.
+        // However, due to the current AppendVec implementation, accounts may remain in the index after
+        // being wiped and reinitialized as something else. We include the redundant filters here to
+        // avoid returning these accounts.
+        //
+        // Filter on Token Account state
+        filters.push(RpcFilterType::DataSize(
+            TokenAccount::get_packed_len() as u64
+        ));
+        // Filter on Mint address
+        filters.push(RpcFilterType::Memcmp(Memcmp {
+            offset: 0,
+            bytes: MemcmpEncodedBytes::Binary(mint_key.to_string()),
+            encoding: None,
+        }));
+        if self.config.supported_indexes.contains(&IndexType::Mint) {
+            bank.get_filtered_indexed_accounts(&IndexKey::Mint(*mint_key), |account| {
+                filters.iter().all(|filter_type| match filter_type {
+                    RpcFilterType::DataSize(size) => account.data.len() as u64 == *size,
+                    RpcFilterType::Memcmp(compare) => compare.bytes_match(&account.data),
```

Comment:
Well, bytes_match does `bs58::decode()`.. Meaning we waste cpu cycles by rather heavy decoding for the number of returned accounts? I wonder how this could be next bottoleneck after the introduction of indexes..

If not too much hassle, could you decode when we construct `RpcFilterType::Memcmp`?

Admittedly, that's a bit unfortunate we must do this extra filter according to this comment: https://github.com/solana-labs/solana/pull/14212/files#diff-cf429ddb403b689f476ba2ee33dc2f4b8a509e93f5b5f901f38874a21cc65959R1366

Reply:
Also, this could improve DX abit. That's because we're squashing base58 decoding errors inside `bytes_match` into brutal `false` (not hit)..

Reply:
That's a really nice idea! I think I'd prefer to do that in a small follow-up PR